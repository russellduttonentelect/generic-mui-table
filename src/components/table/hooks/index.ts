export * from './useSelectedRows';
export * from './useTableQueryParams';
