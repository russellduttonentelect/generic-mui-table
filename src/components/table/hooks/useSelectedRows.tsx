import { useState } from 'react';

export const useSelectedRows = (defaultSelection?: string[]) => {
  const [selected, setSelected] = useState(defaultSelection ?? []);

  const updateRowSelection = (ids: string | string[]) => {
    if (Array.isArray(ids)) {
      setSelected(ids);
      return;
    }

    const selectedIndex = selected.indexOf(ids);
    let newSelected: string[] = [];

    if (selectedIndex === -1) {
      newSelected = newSelected.concat(selected, ids);
    } else if (selectedIndex === 0) {
      newSelected = newSelected.concat(selected.slice(1));
    } else if (selectedIndex === selected.length - 1) {
      newSelected = newSelected.concat(selected.slice(0, -1));
    } else if (selectedIndex > 0) {
      newSelected = newSelected.concat(
        selected.slice(0, selectedIndex),
        selected.slice(selectedIndex + 1)
      );
    }

    setSelected(newSelected);
  };

  const checkIfSelected = (id: string) => selected.includes(id);

  return { selected, updateRowSelection, checkIfSelected };
};
