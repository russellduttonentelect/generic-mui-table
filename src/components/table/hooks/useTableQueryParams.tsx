import { SortDirection } from '@mui/material';
import {
  useQueryParams,
  withDefault,
  StringParam,
  NumberParam
} from 'use-query-params';

const OrderParam = {
  encode: (order: SortDirection) => order as string,
  decode: (
    order: string | null | undefined | (string | null)[]
  ): SortDirection =>
    typeof order === 'string' ? (order as SortDirection) : 'asc'
};

interface DefaultQueryParams {
  defaultPageSize: number;
  disableSort?: boolean;
}

export const useTableQueryParams = ({
  defaultPageSize,
  disableSort = false
}: DefaultQueryParams) => {
  const [query, setQuery] = useQueryParams({
    order: OrderParam,
    orderBy: withDefault(StringParam, ''),
    page: withDefault(NumberParam, 0),
    pageSize: withDefault(NumberParam, defaultPageSize),
    search: StringParam
  });

  const { order, orderBy, page, pageSize, search } = query;

  const setOrder = (order: SortDirection) => {
    if (!disableSort) {
      setQuery({ order }, 'replaceIn');
    }
  };

  const setOrderBy = (orderBy: string) => {
    if (!disableSort) {
      setQuery({ orderBy }, 'replaceIn');
    }
  };

  const setPage = (page: number) => setQuery({ page }, 'replaceIn');
  const setPageSize = (pageSize: number) => setQuery({ pageSize }, 'replaceIn');

  const handlePageChange = (page: number) => {
    setPage(page);
  };

  const handlePageSizeChange = (pageSize: number) => {
    setPageSize(pageSize);
    setPage(0);
  };

  const setSearch = (search: string) => setQuery({ search }, 'replaceIn');

  return {
    order,
    setOrder,
    orderBy,
    setOrderBy,
    page,
    handlePageChange,
    pageSize,
    handlePageSizeChange,
    search,
    setSearch
  };
};
