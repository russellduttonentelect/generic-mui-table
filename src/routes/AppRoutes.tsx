import { AppLayout } from 'components/AppLayout';
import { Home, NotFound } from 'pages';
import { Route, Routes } from 'react-router-dom';

export const AppRoutes = () => {
  return (
    <Routes>
      <Route path='/' element={<AppLayout />}>
        <Route index element={<Home />} />
      </Route>
      <Route path='*' element={<NotFound />} />
    </Routes>
  );
};
