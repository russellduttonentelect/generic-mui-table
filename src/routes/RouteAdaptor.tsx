import { FC, useMemo } from 'react';
import { Location, useLocation, useNavigate } from 'react-router-dom';

export const RouteAdaptor: FC = ({ children }) => {
  const navigate = useNavigate();
  const location = useLocation();

  const adaptedHistory = useMemo(
    () => ({
      replace: (location: Location) => navigate(location, { replace: true, state: location.state }),
      push: (location: Location) => navigate(location, { replace: false, state: location.state })
    }),
    [navigate]
  );

  // https://github.com/pbeshai/use-query-params/issues/196
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return children({ history: adaptedHistory, location });
};
