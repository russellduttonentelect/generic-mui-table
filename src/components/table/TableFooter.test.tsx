import { render } from '@testing-library/react';
import { TableFooter } from '.';
import { useTableQueryParams } from './hooks';
import { QueryParamProvider } from 'use-query-params';
import { RouteAdaptor } from 'src/common/routes/RouteAdaptor';
import { MemoryRouter } from 'react-router-dom';

describe('Table Footer tests', () => {
  test('should render the table footer correctly with sensible parameters', () => {
    const Footer = () => {
      const { page, handlePageChange, pageSize, handlePageSizeChange } =
        useTableQueryParams({ defaultPageSize: 10 });
      return (
        <TableFooter
          handlePageChange={handlePageChange}
          handlePageSizeChange={handlePageSizeChange}
          page={page}
          pageSize={pageSize}
          rowCount={50}
        />
      );
    };

    const { baseElement } = render(
      <MemoryRouter>
        <QueryParamProvider ReactRouterRoute={RouteAdaptor}>
          <Footer />
        </QueryParamProvider>
      </MemoryRouter>
    );

    expect(baseElement).toMatchSnapshot();
  });
});
