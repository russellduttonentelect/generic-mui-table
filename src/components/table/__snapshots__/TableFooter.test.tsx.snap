// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Footer tests should render the table footer correctly with sensible parameters 1`] = `
<body>
  <div>
    <div
      class="MuiTablePagination-root css-jtlhu6-MuiTablePagination-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-11bfvty-MuiToolbar-root-MuiTablePagination-toolbar"
      >
        <div
          class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
        />
        <p
          class="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
          id="mui-2"
        >
          Rows per page:
        </p>
        <div
          class="MuiInputBase-root MuiInputBase-colorPrimary css-16c50h-MuiInputBase-root-MuiTablePagination-select"
          variant="standard"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="mui-2 mui-1"
            class="MuiTablePagination-select MuiSelect-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
            id="mui-1"
            role="button"
            tabindex="0"
          >
            10
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value="10"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTablePagination-selectIcon MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
        <p
          class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
        >
          1â€“10 of 50
        </p>
        <div
          class="MuiTablePagination-actions"
        >
          <button
            aria-label="Go to first page"
            class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            disabled=""
            tabindex="-1"
            title="Go to first page"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FirstPageIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
              />
            </svg>
          </button>
          <button
            aria-label="Go to previous page"
            class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
            disabled=""
            tabindex="-1"
            title="Go to previous page"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowLeftIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </button>
          <button
            aria-label="Go to next page"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            title="Go to next page"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowRightIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Go to last page"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            title="Go to last page"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="LastPageIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
